docs/proposals/20231122-automate-aks-features.md

```diff
@@ -0,0 +1,359 @@
+---
+title: Automate AKS Features Available in CAPZ
+authors:
+  - "@nojnhuh"
+reviewers:
+  - "@CecileRobertMichon"
+  - "@matthchr"
+  - "@dtzar"
+  - "@mtougeron"
+creation-date: 2023-11-22
+last-updated: 2023-11-28
+status: provisional
+see-also:
+  - "docs/proposals/20230123-azure-service-operator.md"
+---
+
+# Automate AKS Features Available in CAPZ
+
+## Table of Contents
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals/Future Work](#non-goalsfuture-work)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+    - [Story 3](#story-3)
+  - [API Design Options](#api-design-options)
+    - [Option 1: CAPZ resource references an existing ASO resource](#option-1-capz-resource-references-an-existing-aso-resource)
+    - [Option 2: CAPZ resource references a non-functional ASO "template" resource](#option-2-capz-resource-references-a-non-functional-aso-template-resource)
+    - [Option 3: CAPZ resource defines an entire unstructured ASO resource inline](#option-3-capz-resource-defines-an-entire-unstructured-aso-resource-inline)
+    - [Option 4: CAPZ resource defines an entire typed ASO resource inline](#option-4-capz-resource-defines-an-entire-typed-aso-resource-inline)
+    - [Option 5: No change: CAPZ resource evolution proceeds the way it currently does](#option-5-no-change-capz-resource-evolution-proceeds-the-way-it-currently-does)
```

Comment:
Is there any way we can consider hybrid approach options where the user uses the existing CAPZ definitions as-is and then _only_ use ASO native definition to supplement what is missing for that same cluster?

Reply:
That should be possible and I think that's actually what the proposal described before today:
https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/0535437c37f2e6e72d165332957b06b07b68897d/docs/proposals/20231122-automate-aks-features.md#upgrade-strategy
> The new API field allowing the entire ASO resource to be defined in the CAPZ API will overlap with several
existing CAPZ API fields. e.g. with [Option 4], AzureManagedControlPlane's `spec.location` maps to the same
ASO API field as `spec.managedClusterSpec.location`. If both are defined, then the validating webhook will
reject that and instruct the user to prefer the `spec.managedCluster.*` field. If only `spec.location` is
defined, its value will still be used to construct the desired state for the ManagedCluster.

Do you envision keeping that split between the CAPZ portion and the ASO definition moving forward, or use that as a transition towards using solely the ASO definition? I think aiming to settle on one consistent set of fields makes the most sense.

Reply:
I'm comfortable with a long/permanent-tail of let's say 80% use-case covered natively in CAPZ and the 20% coverage with the ASO definition to fill the gap.  Ideally, we could have telemetry to know what ASO fields people are having to put in and then prioritize those ones to keep the ~80% target.

Reply:
I believe the long-term path should NOT be a shell CAPZ wrapper around 100% ASO code definition.  For this use case, I feel like people might as well just use ASO by itself as the CAPZ value-add is minimized.